hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieBaseRelation.scala

```diff
@@ -333,7 +335,13 @@ object HoodieBaseRelation {
     partitionedFile => {
       val extension = FSUtils.getFileExtension(partitionedFile.filePath)
       if (HoodieFileFormat.PARQUET.getFileExtension.equals(extension)) {
-        parquetReader.apply(partitionedFile)
+        val iter = parquetReader.apply(partitionedFile)
+        if (iter.isInstanceOf[Closeable]) {
+          // register a callback to close parquetReader which will be executed on task completion.
+          // when tasks finished, this method will be called, and release resources.
+          Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => iter.asInstanceOf[Closeable].close()))
```

Comment:
While i appreciate the intent here to tie up the iterator to the scope of particular task, i don't think this is the right way to fix it: you're tying the lifespan of the iterator to that one of the task (which in this case runs on executor), but there's no clear invariant why this iterator could not outlive this task.

Instead we should rely on the RDD to close out the iterator when its done with iteration. And if you would take a look at `FileScanRDD` (which we rely on) you can see that it does exactly that. The reason why it's broken right now is b/c we modify the iterator (which is not inheriting from Closeable anymore):

```
file: PartitionedFile => {
      val iter = readParquetFile(file)
      iter.flatMap {
        case r: InternalRow => Seq(r)
        case b: ColumnarBatch => b.rowIterator().asScala
      }
    }
```

Reply:
synced up offline. there is no follow up here. things are ok as it is.

Reply:
@nsivabalan i don't think that was the conclusion -- we concluded that we're okay to use `TaskContext` to close the reader (since Spark 2 isn't properly closing readers itself), but we should unify this logic w/in `HoodieBaseRelation`